internal/unionstore/art/art_node.go

```diff
@@ -12,47 +12,663 @@
+	return a.getNode48(an.addr)
+}
+
+func (an *artNode) asNode256(a *artAllocator) *node256 {
+	return a.getNode256(an.addr)
+}
+
+func longestCommonPrefix(l1Key, l2Key artKey, depth uint32) uint32 {
+	idx, limit := depth, min(uint32(len(l1Key)), uint32(len(l2Key)))
+	// TODO: possible optimization
+	// Compare the key by loop can be very slow if the final LCP is large.
+	// Maybe optimize it by comparing the key in chunks if the limit exceeds certain threshold.
+	for ; idx < limit; idx++ {
+		if l1Key[idx] != l2Key[idx] {
+			break
+		}
+	}
+	return idx - depth
+}
+
+// Find the minimum artLeaf under an artNode
+func minimum(a *artAllocator, an artNode) artNode {
+	for {
+		switch an.kind {
+		case typeLeaf:
+			return an
+		case typeNode4:
+			n4 := an.asNode4(a)
+			if !n4.inplaceLeaf.addr.IsNull() {
+				return n4.inplaceLeaf
+			}
+			an = n4.children[0]
+		case typeNode16:
+			n16 := an.asNode16(a)
+			if !n16.inplaceLeaf.addr.IsNull() {
+				return n16.inplaceLeaf
+			}
+			an = n16.children[0]
+		case typeNode48:
+			n48 := an.asNode48(a)
+			if !n48.inplaceLeaf.addr.IsNull() {
+				return n48.inplaceLeaf
+			}
+			idx := n48.nextPresentIdx(0)
+			an = n48.children[n48.keys[idx]]
+		case typeNode256:
+			n256 := an.asNode256(a)
+			if !n256.inplaceLeaf.addr.IsNull() {
+				return n256.inplaceLeaf
+			}
+			idx := n256.nextPresentIdx(0)
+			an = n256.children[idx]
+		case typeInvalid:
+			return nullArtNode
+		}
+	}
+}
+
+func (an *artNode) findChild(a *artAllocator, c byte, valid bool) (int, artNode) {
+	if !valid {
+		return inplaceIndex, an.asNode(a).inplaceLeaf
+	}
+	switch an.kind {
+	case typeNode4:
+		return an.findChild4(a, c)
+	case typeNode16:
+		return an.findChild16(a, c)
+	case typeNode48:
+		return an.findChild48(a, c)
+	case typeNode256:
+		return an.findChild256(a, c)
+	}
+	return notExistIndex, nullArtNode
+}
+
+func (an *artNode) findChild4(a *artAllocator, c byte) (int, artNode) {
+	n4 := an.asNode4(a)
+	for idx := 0; idx < int(n4.nodeNum); idx++ {
+		if n4.keys[idx] == c {
+			return idx, n4.children[idx]
+		}
+	}
+
+	return -2, nullArtNode
+}
+
+func (an *artNode) findChild16(a *artAllocator, c byte) (int, artNode) {
+	n16 := an.asNode16(a)
+
+	idx, found := sort.Find(int(n16.nodeNum), func(i int) int {
+		if n16.keys[i] < c {
+			return 1
+		}
+		if n16.keys[i] == c {
+			return 0
+		}
+		return -1
+	})
+	if found {
+		return idx, n16.children[idx]
+	}
+	return -2, nullArtNode
+}
+
+func (an *artNode) findChild48(a *artAllocator, c byte) (int, artNode) {
+	n48 := an.asNode48(a)
+	if n48.present[c>>n48s]&(1<<(c%n48m)) != 0 {
+		return int(c), n48.children[n48.keys[c]]
+	}
+	return -2, nullArtNode
+}
+
+func (an *artNode) findChild256(a *artAllocator, c byte) (int, artNode) {
+	n256 := an.asNode256(a)
+	if n256.present[c>>n48s]&(1<<(c%n48m)) != 0 {
+		return int(c), n256.children[c]
+	}
+	return -2, nullArtNode
+}
+
+func (an *artNode) swapChild(a *artAllocator, c byte, child artNode) {
```

Comment:
There seems no `swapping` operation, but `replacing` the child pointer, how about naming it `replacingChild`?